pipeline_utils/lib/yaml_parser.py

```diff
@@ -0,0 +1,473 @@
+from pipeline_utils.schemas.yaml_workflow import yaml_workflow_schema
+from pipeline_utils.schemas.yaml_metaworkflow import yaml_metaworkflow_schema
+from pipeline_utils.schemas.yaml_software import yaml_software_schema
+from pipeline_utils.schemas.yaml_file_reference import yaml_file_reference_schema
+from pipeline_utils.schemas.yaml_file_format import yaml_file_format_schema
+
+
+###############################################################
+#   Logger
+###############################################################
+logger = structlog.getLogger(__name__)
+
+
+###############################################################
+#   Functions
+###############################################################
+def load_yaml(file):
+    """Return a generator to YAML documents in file
+    """
+    with open(file) as stream:
+        try:
+            for d in yaml.safe_load_all(stream):
+                yield d
+        except yaml.YAMLError as exc:
+            sys.exit(exc)
+
+def peek(iterable):
+    """
+    """
+    try:
+        first = next(iterable)
+    except StopIteration:
+        return None
+
+    return itertools.chain([first], iterable)
+
+
+###############################################################
+#   SchemaError
+###############################################################
+class SchemaError(Exception):
+    """Custom exception for error tracking
+    """
+
+    def __init__(self):
+        message = 'YAML object failed schema validation'
+        super().__init__(message)
+
+
+###############################################################
+#   YAMLTemplate
+###############################################################
+class YAMLTemplate(object):
+    """
+    """
+
+    # Schema constants
+    NAME_SCHEMA = 'name'
+    TITLE_SCHEMA = 'title'
+    DESCRIPTION_SCHEMA = 'description'
+    ALIASES_SCHEMA = 'aliases'
+    PROJECT_SCHEMA = 'project'
+    INSTITUTION_SCHEMA = 'institution'
+    VERSION_SCHEMA = 'version'
+    ACCESSION_SCHEMA = 'accession'
+    UUID_SCHEMA = 'uuid'
+    ARGUMENT_TYPE_SCHEMA = 'argument_type'
+    ARGUMENT_FORMAT_SCHEMA = 'argument_format'
+    ARGUMENT_NAME_SCHEMA = 'argument_name'
+    VALUE_TYPE_SCHEMA = 'value_type'
+    WORKFLOW_ARGUMENT_NAME_SCHEMA = 'workflow_argument_name'
+    INPUT_SCHEMA = 'input'
+    STATUS_SCHEMA = 'status'
+    SECONDARY_FILES_SCHEMA = 'secondary_files'
+    SECONDARY_FORMATS_SCHEMA = 'secondary_formats'
+    FILE_FORMAT_SCHEMA = 'file_format'
+    SECONDARY_FILE_FORMATS_SCHEMA = 'secondary_file_formats'
+    INSTITUTIONS_SCHEMA = 'institutions'
+    PROJECTS_SCHEMA = 'projects'
+
+    def __init__(self, data, schema):
+        """
+        """
+        self.data = data
+        self.schema = schema
+
+    def _validate(self):
+        """
+        """
+        draft202012validator = Draft202012Validator(self.schema)
+        errors = draft202012validator.iter_errors(self.data)
+        errors_ = peek(errors)
```

Comment:
"Peeking" works here. An alternative would be to just make it a list since I don't think there's much benefit to keeping the errors within a generator here as you'll be going through all of them and memory isn't really a concern. Probably faster with a generator, but likely not that important.
```suggestion
        errors = list(draft202012validator.iter_errors(self.data))
```